### PR TITLE
Verified in the ubuntu environment and updated manual build procedures.

### PR DESCRIPTION
**Pull request by ryo-opening-line**
Verified docs/BUILD-manual.md in my ubuntu environment and updated the manual.

Please see the following my article(JP) showing the building steps in my ubuntu environment. 
https://qiita.com/ryosasa/items/a7404b3df1dd665fd823

### CODE REVIEW

File: docs/BUILD-manual.md [hunk 1]
```diff
@@ -165,7 +168,10 @@ cd catapult-server
 mkdir _build && cd _build
 cmake BOOST_ROOT="${CAT_DEPS_DIR}/boost" cmake .. \
 	-DCMAKE_BUILD_TYPE=RelWithDebInfo \
-	-DCMAKE_PREFIX_PATH="${CAT_DEPS_DIR}/facebook:${CAT_DEPS_DIR}/google:${CAT_DEPS_DIR}/mongodb:${CAT_DEPS_DIR}/zeromq" \
-	\
+	-DCMAKE_PREFIX_PATH="${CAT_DEPS_DIR}/facebook;${CAT_DEPS_DIR}/google;${CAT_DEPS_DIR}/mongodb;${CAT_DEPS_DIR}/zeromq" \
+	-DBoost_INCLUDE_DIR=${CAT_DEPS_DIR}/boost/include \
```
Comment (Jaguar0625) [inline docs/BUILD-manual.md:172]: Are you sure this is needed? We thought setting `BOOST_ROOT` should be sufficient?

File: docs/BUILD-manual.md [hunk 2]
```diff
@@ -89,7 +92,7 @@ function install_google_test {
 
 function install_google_benchmark {
 	cmake_options=()
-	cmake_options+=(-DBENCHMARK_ENABLE_GTEST_TESTS=OFF)
+	cmake_options+=(-DBENCHMARK_ENABLE_GTEST_TESTS=OFF -DCMAKE_POSITION_INDEPENDENT_CODE=ON)
```
Comment (Jaguar0625) [inline docs/BUILD-manual.md:95]: Are you sure this is required?

File: docs/BUILD-manual.md [hunk 3]
```diff
@@ -16,6 +16,9 @@ Required
 Recommended
 
  - ninja-build
+ - zsh
+ - pkg-config
+ - build-essential
```
Comment (Jaguar0625) [inline docs/BUILD-manual.md:21]: Why is `build-essential` recommended?

File: docs/BUILD-manual.md (hunk 1 above)
Comment (ryo-opening-line) [reply docs/BUILD-manual.md:172]: thanks for your review, @Jaguar0625 

yes. Boost_INCLUDE_DIR is needed on my end(ubuntu on Amazon EC2).
I got the following error without -DBoost_INCLUDE_DIR when running cmake.

----
% cmake BOOST_ROOT="${CAT_DEPS_DIR}/boost" cmake .. \
    -DCMAKE_BUILD_TYPE=RelWithDebInfo \
    -DCMAKE_PREFIX_PATH="${CAT_DEPS_DIR}/facebook;${CAT_DEPS_DIR}/google;${CAT_DEPS_DIR}/mongodb;${CAT_DEPS_DIR}/zeromq" \
    \
    -GNinja
-- The C compiler identification is GNU 9.3.0
-- The CXX compiler identification is Clang 10.0.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found GTest: /home/ubuntu/cat_deps_dir/google/lib/libgtest.a  
-- CATAPULT_VERSION_DESCRIPTION: 6d2fc75f [main]
-- USE_CONAN:                    
-- CMAKE_CXX_COMPILER_ID: Clang
-- CMAKE_CXX_FLAGS:                     -stdlib=libc++          -Weverything            -Werror                 -Wno-c++98-compat             -Wno-c++98-compat-pedantic              -Wno-disabled-macro-expansion           -Wno-padded             -Wno-switch-enum              -Wno-weak-vtables -fvisibility=hidden
-- CMAKE_PREFIX_PATH:     /home/ubuntu/cat_deps_dir/facebook;/home/ubuntu/cat_deps_dir/google;/home/ubuntu/cat_deps_dir/mongodb;/home/ubuntu/cat_deps_dir/zeromq
-- CMAKE_MODULE_PATH:     
-- CMAKE_BUILD_TYPE:      RelWithDebInfo
--- locating boost dependencies ---
-- Found Threads: TRUE  
CMake Error at /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:146 (message):
  Could NOT find Boost (missing: Boost_INCLUDE_DIR atomic chrono date_time
  filesystem log log_setup program_options regex thread)
Call Stack (most recent call first):
  /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:393 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.16/Modules/FindBoost.cmake:2179 (find_package_handle_standard_args)
  CMakeLists.txt:33 (find_package)


-- Configuring incomplete, errors occurred!
See also "/home/ubuntu/catapult-server/_build/CMakeFiles/CMakeOutput.log".
See also "/home/ubuntu/catapult-server/_build/CMakeFiles/CMakeError.log".
----

File: docs/BUILD-manual.md (hunk 3 above)
Comment (ryo-opening-line) [reply docs/BUILD-manual.md:21]: thanks for your review, @Jaguar0625 .

I got the following error without build-essential when running ninja.

----
% ninja -j4
[1/2111] Building CXX object src/catapult/api/CMakeFiles/catapult.api.dir/RemoteTransactionApi.cpp.o
FAILED: src/catapult/api/CMakeFiles/catapult.api.dir/RemoteTransactionApi.cpp.o 
/usr/bin/c++  -DBOOST_ALL_DYN_LINK -DBOOST_ASIO_USE_TS_EXECUTOR_AS_DEFAULT -DOPENSSL_API_COMPAT=0x10100000L -I../ -I../src -isystem /home/ubuntu/cat_deps_dir/boost/include -stdlib=libc++            -Weverything            -Werror                 -Wno-c++98-compat             -Wno-c++98-compat-pedantic              -Wno-disabled-macro-expansion           -Wno-padded          -Wno-switch-enum                 -Wno-weak-vtables -fvisibility=hidden -O2 -g -DNDEBUG -g1 -fPIC   -std=gnu++17 -MD -MT src/catapult/api/CMakeFiles/catapult.api.dir/RemoteTransactionApi.cpp.o -MF src/catapult/api/CMakeFiles/catapult.api.dir/RemoteTransactionApi.cpp.o.d -o src/catapult/api/CMakeFiles/catapult.api.dir/RemoteTransactionApi.cpp.o -c ../src/catapult/api/RemoteTransactionApi.cpp
In file included from ../src/catapult/api/RemoteTransactionApi.cpp:22:
In file included from ../src/catapult/api/RemoteTransactionApi.h:23:
In file included from ../src/catapult/api/RemoteApi.h:23:
In file included from ../src/catapult/model/NodeIdentity.h:23:
In file included from ../src/catapult/types.h:23:
In file included from ../src/catapult/utils/ByteArray.h:23:
In file included from ../src/catapult/utils/HexFormatter.h:23:
../src/catapult/utils/StreamFormatGuard.h:24:10: fatal error: 'iomanip' file not found
#include <iomanip>
         ^~~~~~~~~
1 error generated.
[2/2111] Building CXX object src/catapult/api/CMakeFiles/catapult.api.dir/RemoteChainApi.cpp.o
FAILED: src/catapult/api/CMakeFiles/catapult.api.dir/RemoteChainApi.cpp.o 
/usr/bin/c++  -DBOOST_ALL_DYN_LINK -DBOOST_ASIO_USE_TS_EXECUTOR_AS_DEFAULT -DOPENSSL_API_COMPAT=0x10100000L -I../ -I../src -isystem /home/ubuntu/cat_deps_dir/boost/include -stdlib=libc++            -Weverything            -Werror                 -Wno-c++98-compat             -Wno-c++98-compat-pedantic              -Wno-disabled-macro-expansion           -Wno-padded          -Wno-switch-enum                 -Wno-weak-vtables -fvisibility=hidden -O2 -g -DNDEBUG -g1 -fPIC   -std=gnu++17 -MD -MT src/catapult/api/CMakeFiles/catapult.api.dir/RemoteChainApi.cpp.o -MF src/catapult/api/CMakeFiles/catapult.api.dir/RemoteChainApi.cpp.o.d -o src/catapult/api/CMakeFiles/catapult.api.dir/RemoteChainApi.cpp.o -c ../src/catapult/api/RemoteChainApi.cpp
In file included from ../src/catapult/api/RemoteChainApi.cpp:22:
In file included from ../src/catapult/api/RemoteChainApi.h:23:
In file included from ../src/catapult/api/ChainApi.h:23:
In file included from ../src/catapult/api/ApiTypes.h:23:
In file included from ../src/catapult/exceptions.h:24:
../src/catapult/utils/BaseValue.h:24:10: fatal error: 'iosfwd' file not found
#include <iosfwd>
         ^~~~~~~~
1 error generated.
[3/2111] Building CXX object src/catapult/api/CMakeFiles/catapult.api.dir/LocalChainApi.cpp.o
FAILED: src/catapult/api/CMakeFiles/catapult.api.dir/LocalChainApi.cpp.o 
/usr/bin/c++  -DBOOST_ALL_DYN_LINK -DBOOST_ASIO_USE_TS_EXECUTOR_AS_DEFAULT -DOPENSSL_API_COMPAT=0x10100000L -I../ -I../src -isystem /home/ubuntu/cat_deps_dir/boost/include -stdlib=libc++            -Weverything            -Werror                 -Wno-c++98-compat             -Wno-c++98-compat-pedantic              -Wno-disabled-macro-expansion           -Wno-padded          -Wno-switch-enum                 -Wno-weak-vtables -fvisibility=hidden -O2 -g -DNDEBUG -g1 -fPIC   -std=gnu++17 -MD -MT src/catapult/api/CMakeFiles/catapult.api.dir/LocalChainApi.cpp.o -MF src/catapult/api/CMakeFiles/catapult.api.dir/LocalChainApi.cpp.o.d -o src/catapult/api/CMakeFiles/catapult.api.dir/LocalChainApi.cpp.o -c ../src/catapult/api/LocalChainApi.cpp
In file included from ../src/catapult/api/LocalChainApi.cpp:22:
In file included from ../src/catapult/api/LocalChainApi.h:23:
In file included from ../src/catapult/api/ChainApi.h:23:
In file included from ../src/catapult/api/ApiTypes.h:23:
In file included from ../src/catapult/exceptions.h:24:
../src/catapult/utils/BaseValue.h:24:10: fatal error: 'iosfwd' file not found
#include <iosfwd>
         ^~~~~~~~
1 error generated.
[4/2111] Building C object external/CMakeFiles/external.dir/donna/ed25519.c.o
ninja: build stopped: subcommand failed.
----

File: docs/BUILD-manual.md (hunk 2 above)
Comment (ryo-opening-line) [reply docs/BUILD-manual.md:95]: thanks for your review, @Jaguar0625 .

yes. I got the following error without -DCMAKE_POSITION_INDEPENDENT_CODE=ON in install_google_benchmark when running ninja.
```
% ninja -j4
...
[315/2111] Linking CXX executable bin/bench.catapult.crypto.hashers
FAILED: bin/bench.catapult.crypto.hashers 
: && /usr/bin/c++  -Wall -Wextra -Wpedantic -Wshadow -Wconversion -Wformat-security -Werror -Wstrict-aliasing=1 -fvisibility=hidden -O2 -g -DNDEBUG -Wno-maybe-uninitialized -g1 -fno-omit-frame-pointer   src/catapult/version/nix/CMakeFiles/catapult.version.nix.dir/what_version.cpp.o tests/bench/crypto/hashers/CMakeFiles/bench.catapult.crypto.hashers.dir/HashersBench.cpp.o  -o bin/bench.catapult.crypto.hashers  /home/ubuntu/cat_deps_dir/google/lib/libbenchmark.a  /home/ubuntu/cat_deps_dir/boost/lib/libboost_atomic.so  /home/ubuntu/cat_deps_dir/boost/lib/libboost_chrono.so  /home/ubuntu/cat_deps_dir/boost/lib/libboost_date_time.so  /home/ubuntu/cat_deps_dir/boost/lib/libboost_filesystem.so  /home/ubuntu/cat_deps_dir/boost/lib/libboost_log.so  /home/ubuntu/cat_deps_dir/boost/lib/libboost_log_setup.so  /home/ubuntu/cat_deps_dir/boost/lib/libboost_program_options.so  /home/ubuntu/cat_deps_dir/boost/lib/libboost_regex.so  /home/ubuntu/cat_deps_dir/boost/lib/libboost_thread.so  -lpthread  lib/libcatapult.crypto.a  lib/libbench.catapult.bench.nodeps.a  lib/libexternal.a  /usr/lib/x86_64-linux-gnu/libssl.so  /usr/lib/x86_64-linux-gnu/libcrypto.so  lib/libcatapult.utils.a  /home/ubuntu/cat_deps_dir/google/lib/libbenchmark.a  -pthread  /usr/lib/x86_64-linux-gnu/librt.so  lib/libcatapult.version.a  /home/ubuntu/cat_deps_dir/boost/lib/libboost_atomic.so  /home/ubuntu/cat_deps_dir/boost/lib/libboost_chrono.so  /home/ubuntu/cat_deps_dir/boost/lib/libboost_date_time.so  /home/ubuntu/cat_deps_dir/boost/lib/libboost_filesystem.so  /home/ubuntu/cat_deps_dir/boost/lib/libboost_log.so  /home/ubuntu/cat_deps_dir/boost/lib/libboost_log_setup.so  /home/ubuntu/cat_deps_dir/boost/lib/libboost_program_options.so  /home/ubuntu/cat_deps_dir/boost/lib/libboost_regex.so  /home/ubuntu/cat_deps_dir/boost/lib/libboost_thread.so  -lpthread && :
/usr/bin/ld: /home/ubuntu/cat_deps_dir/google/lib/libbenchmark.a(benchmark.cc.o): relocation R_X86_64_32 against `.bss' can not be used when making a PIE object; recompile with -fPIE
/usr/bin/ld: /home/ubuntu/cat_deps_dir/google/lib/libbenchmark.a(benchmark_name.cc.o): relocation R_X86_64_32 against symbol `__gxx_personality_v0@@CXXABI_1.3' can not be used when making a PIE object; recompile with -fPIE
/usr/bin/ld: /home/ubuntu/cat_deps_dir/google/lib/libbenchmark.a(benchmark_register.cc.o): relocation R_X86_64_32 against `.bss' can not be used when making a PIE object; recompile with -fPIE
/usr/bin/ld: /home/ubuntu/cat_deps_dir/google/lib/libbenchmark.a(benchmark_runner.cc.o): relocation R_X86_64_32 against symbol `FLAGS_benchmark_min_time' can not be used when making a PIE object; recompile with -fPIE
/usr/bin/ld: /home/ubuntu/cat_deps_dir/google/lib/libbenchmark.a(colorprint.cc.o): relocation R_X86_64_32 against symbol `_ZGVZN9benchmark8internal18GetNullLogInstanceEvE3log' can not be used when making a PIE object; recompile with -fPIE
/usr/bin/ld: /home/ubuntu/cat_deps_dir/google/lib/libbenchmark.a(commandlineflags.cc.o): relocation R_X86_64_32 against `.rodata.str1.1' can not be used when making a PIE object; recompile with -fPIE
/usr/bin/ld: /home/ubuntu/cat_deps_dir/google/lib/libbenchmark.a(complexity.cc.o): relocation R_X86_64_32S against `.rodata' can not be used when making a PIE object; recompile with -fPIE
/usr/bin/ld: /home/ubuntu/cat_deps_dir/google/lib/libbenchmark.a(console_reporter.cc.o): relocation R_X86_64_32S against `.rodata.str1.1' can not be used when making a PIE object; recompile with -fPIE
/usr/bin/ld: /home/ubuntu/cat_deps_dir/google/lib/libbenchmark.a(counter.cc.o): relocation R_X86_64_32 against `.rodata' can not be used when making a PIE object; recompile with -fPIE
/usr/bin/ld: /home/ubuntu/cat_deps_dir/google/lib/libbenchmark.a(csv_reporter.cc.o): relocation R_X86_64_32 against `.rodata.str1.1' can not be used when making a PIE object; recompile with -fPIE
/usr/bin/ld: /home/ubuntu/cat_deps_dir/google/lib/libbenchmark.a(json_reporter.cc.o): relocation R_X86_64_32 against `.rodata.str1.1' can not be used when making a PIE object; recompile with -fPIE
/usr/bin/ld: /home/ubuntu/cat_deps_dir/google/lib/libbenchmark.a(reporter.cc.o): relocation R_X86_64_32S against symbol `_ZTVN9benchmark17BenchmarkReporterE' can not be used when making a PIE object; recompile with -fPIE
/usr/bin/ld: /home/ubuntu/cat_deps_dir/google/lib/libbenchmark.a(statistics.cc.o): relocation R_X86_64_32 against `.rodata.str1.1' can not be used when making a PIE object; recompile with -fPIE
/usr/bin/ld: /home/ubuntu/cat_deps_dir/google/lib/libbenchmark.a(string_util.cc.o): relocation R_X86_64_32S against symbol `_ZTVNSt7__cxx1115basic_stringbufIcSt11char_traitsIcESaIcEEE@@GLIBCXX_3.4.21' can not be used when making a PIE object; recompile with -fPIE
/usr/bin/ld: /home/ubuntu/cat_deps_dir/google/lib/libbenchmark.a(sysinfo.cc.o): relocation R_X86_64_32 against `.bss' can not be used when making a PIE object; recompile with -fPIE
/usr/bin/ld: /home/ubuntu/cat_deps_dir/google/lib/libbenchmark.a(timers.cc.o): relocation R_X86_64_32 against `.rodata.str1.1' can not be used when making a PIE object; recompile with -fPIE
/usr/bin/ld: /home/ubuntu/cat_deps_dir/google/lib/libbenchmark.a(benchmark_api_internal.cc.o): relocation R_X86_64_32 against symbol `__gxx_personality_v0@@CXXABI_1.3' can not be used when making a PIE object; recompile with -fPIE
collect2: error: ld returned 1 exit status
[318/2111] Building CXX object tests/catapult/api/CMakeFiles/tests.catapult.api.dir/RemoteChainApiTests.cpp.o
ninja: build stopped: subcommand failed.
```

**Comment by Jaguar0625** [comment]
Can you share your OS and compiler versions?

I noticed something strange in your log:
```
-- The C compiler identification is GNU 9.3.0
-- The CXX compiler identification is Clang 10.0.0
```

Are you attempting to build with clang or gcc?

File: docs/BUILD-manual.md (hunk 3 above)
Comment (ryo-opening-line) [reply docs/BUILD-manual.md:21]: thanks for confirmation, @Jaguar0625 .

the followings are in my end(ubuntu default setting).
* os: ubuntu 20.04.1
* gcc: 9.3.0
* clang: 10.0.0

```
$ cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.1 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.1 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal

$ gcc --version
gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

$ clang --version
clang version 10.0.0-4ubuntu1 
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

Comment (ryo-opening-line) [reply docs/BUILD-manual.md:21]: Regarding to gcc or clang, I left it up to cmake to decide which compiler to use...
Will confirm it later.

File: docs/BUILD-manual.md (hunk 1 above)
Comment (ryo-opening-line) [reply docs/BUILD-manual.md:172]: @Jaguar0625 
Regarding to Boost_INCLUDE_DIR,  I found a typo in BUILD-manual.md.
It works without Boost_INCLUDE_DIR.

```diff
diff --git a/docs/BUILD-manual.md b/docs/BUILD-manual.md
index af93a3cd..456960c7 100644
--- a/docs/BUILD-manual.md
+++ b/docs/BUILD-manual.md
@@ -160,7 +160,7 @@ git clone https://github.com/nemtech/catapult-server.git
 cd catapult-server
 
 mkdir _build && cd _build
-cmake BOOST_ROOT="${CAT_DEPS_DIR}/boost" cmake .. \
+BOOST_ROOT="${CAT_DEPS_DIR}/boost" cmake .. \
        -DCMAKE_BUILD_TYPE=RelWithDebInfo \
        -DCMAKE_PREFIX_PATH="${CAT_DEPS_DIR}/facebook;${CAT_DEPS_DIR}/google;${CAT_DEPS_DIR}/mongodb;${CAT_DEPS_DIR}/zeromq" \
        \
```

File: docs/BUILD-manual.md (hunk 2 above)
Comment (ryo-opening-line) [reply docs/BUILD-manual.md:95]: Should install_google_benchmark() have the same option as install_google_test() using -DCMAKE_POSITION_INDEPENDENT_CODE?

Without this option, -fPIC option for compiler is disabled in my end.

File: docs/BUILD-manual.md (hunk 3 above)
Comment (ryo-opening-line) [reply docs/BUILD-manual.md:21]: @Jaguar0625 
Resolved. 
I made a successfull build without build-essential and -DCMAKE_POSITION_INDEPENDENT_CODE=ON.
The mixture of gcc/g++ and clang seems to have had an bad effect.
A build using gcc/g++(no clang) results successfully.
Thanks for your help!

Comment (ryo-opening-line) [reply docs/BUILD-manual.md:21]: Below are all the commands I have executed on my end that have resulted in a successful build.

```sh
#
# step0 preparing
#
sudo apt update
sudo apt -y upgrade
sudo apt -y install gcc g++ cmake libssl-dev ninja-build zsh pkg-config

zsh

mkdir -p cat_deps_dir/boost
export CAT_DEPS_DIR=$HOME/cat_deps_dir

#
# step1. no change
#
function download_boost {
	local boost_ver=1_${1}_0
	local boost_ver_dotted=1.${1}.0

	curl -o boost_${boost_ver}.tar.gz -SL https://dl.bintray.com/boostorg/release/${boost_ver_dotted}/source/boost_${boost_ver}.tar.gz
	tar -xzf boost_${boost_ver}.tar.gz
	mv boost_${boost_ver} boost
}

function download_git_dependency {
	git clone git://github.com/${1}/${2}.git
	cd ${2}
	git checkout ${3}
	cd ..
}

function download_all {
	download_boost 74

	download_git_dependency google googletest release-1.10.0
	download_git_dependency google benchmark v1.5.2

	download_git_dependency mongodb mongo-c-driver 1.17.2
	download_git_dependency mongodb mongo-cxx-driver r3.6.1

	download_git_dependency zeromq libzmq v4.3.3
	download_git_dependency zeromq cppzmq v4.7.1

	download_git_dependency facebook rocksdb v6.13.3
}

cd ${CAT_DEPS_DIR}
mkdir source
cd source
download_all

#
# step2. no change.
#
boost_output_dir=${CAT_DEPS_DIR}/boost

function install_boost {
	cd boost
	./bootstrap.sh with-toolset=clang --prefix=${boost_output_dir}

	b2_options=()
	b2_options+=(--prefix=${boost_output_dir})
	./b2 ${b2_options[@]} -j 8 stage release
	./b2 install ${b2_options[@]}
}

function install_git_dependency {
	cd ${2}
	mkdir _build
	cd _build

	cmake -DCMAKE_BUILD_TYPE=RelWithDebInfo -DCMAKE_INSTALL_PREFIX="${CAT_DEPS_DIR}/${1}" ${cmake_options[@]} ..
	make -j 8 && make install
}

function install_google_test {
	cmake_options=()
	cmake_options+=(-DCMAKE_POSITION_INDEPENDENT_CODE=ON)
	install_git_dependency google googletest
}

function install_google_benchmark {
	cmake_options=()
	cmake_options+=(-DBENCHMARK_ENABLE_GTEST_TESTS=OFF)
	install_git_dependency google benchmark
}

function install_mongo_c_driver {
	cmake_options=()
	cmake_options+=(-DENABLE_AUTOMATIC_INIT_AND_CLEANUP=OFF)
	cmake_options+=(-DENABLE_MONGODB_AWS_AUTH=OFF)
	cmake_options+=(-DENABLE_TESTS=OFF)
	cmake_options+=(-DENABLE_EXAMPLES=OFF)
	cmake_options+=(-DENABLE_SASL=OFF)
	install_git_dependency mongodb mongo-c-driver
}

function install_mongo_cxx_driver {
	cmake_options=()
	cmake_options+=(-DBOOST_ROOT=${boost_output_dir})
	cmake_options+=(-DCMAKE_CXX_STANDARD=17)
	install_git_dependency mongodb mongo-cxx-driver
}

function install_zmq_lib {
	cmake_options=()
	cmake_options+=(-DWITH_TLS=OFF)
	install_git_dependency zeromq libzmq
}

function install_zmq_cpp {
	cmake_options=()
	cmake_options+=(-DCPPZMQ_BUILD_TESTS=OFF)
	install_git_dependency zeromq cppzmq
}

function install_rocks {
	cmake_options=()
	cmake_options+=(-DPORTABLE=1)
	cmake_options+=(-DWITH_TESTS=OFF)
	cmake_options+=(-DWITH_TOOLS=OFF)
	cmake_options+=(-DWITH_BENCHMARK_TOOLS=OFF)
	cmake_options+=(-DWITH_CORE_TOOLS=OFF)
	cmake_options+=(-DWITH_GFLAGS=OFF)
	install_git_dependency facebook rocksdb
}

function install_all {
	declare -a installers=(
		install_boost
		install_google_test
		install_google_benchmark
		install_mongo_c_driver
		install_mongo_cxx_driver
		install_zmq_lib
		install_zmq_cpp
		install_rocks
	)
	for install in "${installers[@]}"
	do
		pushd source > /dev/null
		${install}
		popd > /dev/null
	done
}

cd ${CAT_DEPS_DIR}
install_all


#
# step3. needs the following update.
# -cmake BOOST_ROOT="${CAT_DEPS_DIR}/boost" cmake .. \
# +BOOST_ROOT="${CAT_DEPS_DIR}/boost" cmake .. \
#
git clone https://github.com/nemtech/catapult-server.git
cd catapult-server

mkdir _build && cd _build
BOOST_ROOT="${CAT_DEPS_DIR}/boost" cmake .. \
	-DCMAKE_BUILD_TYPE=RelWithDebInfo \
	-DCMAKE_PREFIX_PATH="${CAT_DEPS_DIR}/facebook;${CAT_DEPS_DIR}/google;${CAT_DEPS_DIR}/mongodb;${CAT_DEPS_DIR}/zeromq" \
	\
	-GNinja
ninja publish
ninja -j4
```

Comment (ryo-opening-line) [reply docs/BUILD-manual.md:21]: @Jaguar0625 
Please make the following corrections and you'll be done.

```diff
diff --git a/docs/BUILD-manual.md b/docs/BUILD-manual.md
index af93a3cd..456960c7 100644
--- a/docs/BUILD-manual.md
+++ b/docs/BUILD-manual.md
@@ -160,7 +160,7 @@ git clone https://github.com/nemtech/catapult-server.git
 cd catapult-server
 
 mkdir _build && cd _build
-cmake BOOST_ROOT="${CAT_DEPS_DIR}/boost" cmake .. \
+BOOST_ROOT="${CAT_DEPS_DIR}/boost" cmake .. \
        -DCMAKE_BUILD_TYPE=RelWithDebInfo \
        -DCMAKE_PREFIX_PATH="${CAT_DEPS_DIR}/facebook;${CAT_DEPS_DIR}/google;${CAT_DEPS_DIR}/mongodb;${CAT_DEPS_DIR}/zeromq" \
        \
```

Comment (Jaguar0625) [reply docs/BUILD-manual.md:21]: Thank you. Please let me know if this change is correct:
https://github.com/nemtech/catapult-server/compare/build/update-docs-again?short_path=3aef6ee#diff-3aef6ee1abee02ed08ef3b8176ac21efe5c6252850d995c0f8bbcbbad560ba79

File: docs/BUILD-manual.md (hunk 2 above)
Comment (Jaguar0625) [reply docs/BUILD-manual.md:95]: :+1:

File: docs/BUILD-manual.md (hunk 1 above)
Comment (Jaguar0625) [reply docs/BUILD-manual.md:172]: :+1:

File: docs/BUILD-manual.md (hunk 3 above)
Comment (Jaguar0625) [reply docs/BUILD-manual.md:21]: Also please let me know if we should add any text about compiler selection?

Comment (ryo-opening-line) [reply docs/BUILD-manual.md:21]: @Jaguar0625
Thank you very much, Jaguar.

That change is perfect!

Thanks again!

**Comment by Jaguar0625** [comment]
Merged in as 7d0116d6a.

Thank you!

**Pull request closed without merging**